sources/api/netdog/src/dns.rs

```diff
@@ -0,0 +1,258 @@
+//! The dns module contains the code necessary to gather DNS settings from config file, DHCP lease
+//! file, or an already-parsed DHCP lease via LeaseInfo struct.  It also contains the code
+//! necessary to write a properly formatted `resolv.conf`.
+use crate::lease::LeaseInfo;
+use crate::RESOLV_CONF;
+use rand::prelude::SliceRandom;
+use rand::thread_rng;
+use serde::Deserialize;
+use snafu::ResultExt;
+use std::collections::BTreeSet;
+use std::fmt::Write;
+use std::fs;
+use std::iter::FromIterator;
+use std::net::IpAddr;
+use std::path::Path;
+
+static DNS_CONFIG: &str = "/etc/netdog.toml";
+
+#[derive(PartialEq, Debug, Deserialize)]
+pub(crate) struct DnsSettings {
+    #[serde(rename = "name-servers")]
+    nameservers: Option<BTreeSet<IpAddr>>,
+    #[serde(rename = "search-list")]
+    search: Option<Vec<String>>,
+}
+
+impl DnsSettings {
+    /// Create a DnsSettings from TOML config file
+    pub(crate) fn from_config() -> Result<Option<Self>> {
+        Self::from_config_impl(DNS_CONFIG)
+    }
+
+    fn from_config_impl<P>(path: P) -> Result<Option<Self>>
+    where
+        P: AsRef<Path>,
+    {
+        // Ensure we don't attempt to load a nonexistent or empty config file.  At boot time, the
+        // config file won't exist because it hasn't been generated by the API yet.  After that,
+        // the file will always exist because it's a configuration file for settings in the API and
+        // will always be generated.  If the DNS settings aren't populated, the file will be empty.
+        // We can assume if the file is empty that the settings don't exist.
+        let path = path.as_ref();
+        let config_exists = if Path::exists(path) {
+            let file_len = fs::metadata(path)
+                .context(error::DnsConfMetaSnafu { path })?
+                .len();
+            file_len != 0
+        } else {
+            false
+        };
+
+        if config_exists {
+            let config_str =
+                fs::read_to_string(path).context(error::DnsConfReadFailedSnafu { path })?;
+            let dns_config =
+                toml::from_str(&config_str).context(error::DnsConfParseSnafu { path })?;
+
+            Ok(Some(dns_config))
+        } else {
+            eprintln!("No DNS configuration exists in {}", DNS_CONFIG);
+            Ok(None)
+        }
+    }
+
+    /// Create a DnsSettings from LeaseInfo struct
+    pub(crate) fn from_lease_info(info: &LeaseInfo) -> Self {
+        // Randomize name server order, for libc implementations like musl that send
+        // queries to the first N servers.
+        let mut dns_servers: Vec<IpAddr> = info.dns_servers.clone().into_iter().collect();
+        dns_servers.shuffle(&mut thread_rng());
```

Comment:
Is this randomization something we have to do, otherwise the resolver won't work? Or does the libc resolver library do the randomization?